File: 2024/Day 14/Python/main.py
"""Advent of Code - 2024 - Day 14"""

import re


def skip(t: int):
    """skips to state at time t"""
    result: list[tuple[int, int]] = []
    for px, py, vx, vy in drones:
        npx = (px + t * vx) % W
        npy = (py + t * vy) % H
        result.append((npx, npy))
    return result


def print_state(positions: list[tuple[int, int]]):
    """prints state"""
    for cy in range(H):
        print("".join("#" if (cx, cy) in positions else " " for cx in range(W)))


def skip_print_state(t: int):
    """print state at time t"""
    print(f"state after {t} seconds")
    print_state(skip(t))


W = 101
H = 103

PATTERN = r"p=(-?\d+),(-?\d+) v=(-?\d+),(-?\d+)"
drones: list[tuple[int, int, int, int]] = []
with open(0, encoding="utf-8") as f:
    for line in f.read().splitlines():
        if m := re.match(PATTERN, line):
            drones.append(
                (
                    int(m.groups()[0]),
                    int(m.groups()[1]),
                    int(m.groups()[2]),
                    int(m.groups()[3]),
                )
            )

MX = (W - 1) // 2
MY = (H - 1) // 2
q1: int = 0
q2: int = 0
q3: int = 0
q4: int = 0
for x, y in skip(100):
    if x < MX and y < MY:
        q1 += 1
    elif x > MX and y < MY:
        q2 += 1
    elif x < MX and y > MY:
        q3 += 1
    elif x > MX and y > MY:
        q4 += 1

print(q1 * q2 * q3 * q4)

# at 30 there is a horizontal bar that repeats every 103(h) steps
# at 89 there is a vertical bar that repeats every 101(w) steps
# they intersect when t = 30 + a*103 = 89 + b*101
# a = (59 + b*101)/103 where 59 + b*101 is a multiple of 103
# therefore a = 80 b = 81
# therefore they intersect at t = 8270

skip_print_state(8270)

File: 2024/Day 14/C++/main.cpp
#include <string>
#include <regex>
#include <iostream>
#include <tuple>
#include <vector>
#include <algorithm>

const std::regex PATTERN("p=(-?\\d+),(-?\\d+) v=(-?\\d+),(-?\\d+)");

const int W = 101;
const int H = 103;
const int MX = W / 2;
const int MY = H / 2;

std::vector<std::pair<int, int>> skip(int t, std::vector<std::tuple<int, int, int, int>> &drones)
{
    std::vector<std::pair<int, int>> results;
    for (auto const &[px, py, vx, vy] : drones)
    {
        results.push_back(std::make_pair((((px + t * vx) % W) + W) % W, (((py + t * vy) % H) + H) % H));
    }
    return results;
}

void print_state(std::vector<std::pair<int, int>> &state)
{
    for (int y = 0; y < H; y++)
    {
        for (int x = 0; x < W; x++)
        {
            if (std::find(state.begin(), state.end(), std::make_pair(x, y)) != state.end())
            {
                std::cout << "#";
            }
            else
            {
                std::cout << " ";
            }
        }
        std::cout << "\n";
    }
}

int main(int argc, char *argv[])
{
    std::string line;
    std::vector<std::tuple<int, int, int, int>> drones;
    while (std::getline(std::cin, line))
    {
        std::smatch m;
        std::regex_match(line, m, PATTERN);
        drones.push_back({stoi(m[1]), stoi(m[2]), stoi(m[3]), stoi(m[4])});
    }

    int q1 = 0;
    int q2 = 0;
    int q3 = 0;
    int q4 = 0;
    for (auto const &[x, y] : skip(100, drones))
    {
        if (x < MX && y < MY)
        {
            q1 += 1;
        }
        else if (x > MX && y < MY)
        {
            q2 += 1;
        }
        else if (x < MX && y > MY)
        {
            q3 += 1;
        }
        else if (x > MX && y > MY)
        {
            q4 += 1;
        }
    }
    std::cout << q1 * q2 * q3 * q4 << "\n";
    std::vector<std::pair<int, int>> state =skip(8270, drones);
    print_state(state);
}